### PR TITLE
Introduce Slack Integration

### PR DESCRIPTION
**Pull request by potsbo**
## WHY

Developers need to map their Slack accounts and GitHub ones.

## WHAT

Add Slack account map ability
- [x] Add Slack service to fetch SlackID(not username)
- [x] Store Slack account information DynamoDB

## TEST
```console
potsbo@Shimpeis-MacBook-Pro-INS-% make && envchain wtd ./bin/developers-account-mapper register potsbo potsbo shimpei
go build -ldflags="-s -w -X \"main.Version=0.1.0\" -X \"main.Revision=ef04633\"" -o bin/developers-account-mapper
user "potsbo:@potsbo:<@shimpei:<CONSEALED>>" added.
potsbo@Shimpeis-MacBook-Pro-INS-% make && envchain wtd ./bin/developers-account-mapper list
make: `bin/developers-account-mapper' is up to date.
potsbo:@potsbo:<@shimpei:<CONSEALED>>
```

## NOTE

- Test is still not implemented, which has to be done

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -2,22 +2,45 @@ package models
...
+		SlackUserId: slackUserId,
 	}
 }
 
+func (u *User) SetSlackUserId(newId string) *User {
+	u.SlackUserId = newId
+	return u
+}
+
+func (u *User) RetrieveSlackUserId() (*User, error) {
```
Comment (dtan4) [inline models/user.go]: `u` is just a pointer, so this function need not to return itself again.

```go
func (u *User) RetrieveSlackUserId() error {
	// ...
	u.SetSlackUserId(nameIdMap[u.SlackUsername])
	return nil
}

File: models/user.go [hunk 2]
```diff
@@ -2,22 +2,45 @@ package models
...
+	nameIdMap, err := services.SlackUserList()
+	if err != nil {
+		return nil, err
+	}
+	u.SetSlackUserId(nameIdMap[u.SlackUsername])
```
Comment (dtan4) [inline models/user.go]:
```go
u.SlackUserId = nameIdMap[u.SlackUsername]
```

File: services/slack_api.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+	data := map[string]interface{}{}
+	dec := json.NewDecoder(resp.Body)
+	dec.Decode(&data)
+	jq := jsonq.NewQuery(data)
+	arr, err := jq.Array("members")
+
+	users := make(map[string]string)
+	for i := 0; i < len(arr); i++ {
+		id, _ := jq.String("members", strconv.Itoa(i), "id")
```
Comment (dtan4) [inline services/slack_api.go]: Please do not discard returned error as possible.

File: models/user.go [hunk 4]
```diff
@@ -2,22 +2,45 @@ package models
...
+	return u, nil
+}
+
 func (u *User) String() string {
-	return fmt.Sprintf("%v:@%v", u.LoginName, u.GitHubUsername)
+	if u.SlackUserId == "" {
+		u.RetrieveSlackUserId()
```
Comment (dtan4) [inline models/user.go]: Please catch error here.

**Review by dtan4 — CHANGES_REQUESTED** [review]

File: models/user.go (hunk 1 above)
Comment (potsbo) [reply models/user.go]: Fixed

File: models/user.go (hunk 2 above)
Comment (potsbo) [reply models/user.go]: Fixed

File: models/user.go (hunk 4 above)
Comment (potsbo) [reply models/user.go]: Rewrite to catch the error, but actually, I don't know how to handle this. Is that ok to return the error string?

File: services/slack_api.go (hunk 3 above)
Comment (potsbo) [reply services/slack_api.go]: Fixed

**Comment by potsbo** [comment]
@dtan4 
Thank you for the review.
And, this is my question: https://github.com/wantedly/developers-account-mapper/pull/7#discussion_r93889173 

### TEST
It works fine
```console
potsbo@magi-INS-% make && envchain wtd ./bin/developers-account-mapper register potsbo potsbo shimpei
make: `bin/developers-account-mapper' is up to date.
user "potsbo:@potsbo:<@shimpei:U2YBNNR24>" added.
```

File: models/user.go (hunk 4 above)
Comment (dtan4) [reply models/user.go]: I think `String()` should return both `string` and `error`.

```go
func (u *User) String() (string, error)
```

Comment (potsbo) [reply models/user.go]: String is implicitly called here
`fmt.Printf("user %q added.\n", user)`
And, this seems to cause
```console
potsbo@magi-INS-% make && envchain wtd ./bin/developers-account-mapper register potsbo potsbo shimpei
go build -ldflags="-s -w -X \"main.Version=0.1.0\" -X \"main.Revision=8c5a955\"" -o bin/developers-account-mapper
# github.com/wantedly/developers-account-mapper/models
../../../../.go/src/github.com/wantedly/developers-account-mapper/models/user.go:40: cannot use nil as type string in return argument
make: *** [bin/developers-account-mapper] Error 2
```

I'm not 100% sure, though, maybe String should not return anything but a `string`.
Or, do you think we should assign like doing `str, err := user.String` and handle it?

**Review by dtan4 — APPROVED** [review]

**Comment by dtan4** [comment]
LGTM 👍

**Comment by potsbo** [comment]
@dtan4 I need comment for https://github.com/wantedly/developers-account-mapper/pull/7#discussion_r93901791
it's failing

File: models/user.go (hunk 4 above)
Comment (dtan4) [reply models/user.go]: > Or, do you think we should assign like doing str, err := user.String and handle it?

You should do so, or remove calling `RetrieveSlackUserId` in `String()`.

**Comment by dtan4** [comment]
Sorry, I missed that comment 😢

**Pull request merged**